File: ai/agent.ts
/**
 * AI Agent implementation using Vercel AI SDK.
 *
 * Provides streaming and non-streaming interfaces for running Claude-powered agents.
 */

import { anthropic } from '@ai-sdk/anthropic';
// Use explicit path to avoid conflict with local ai/ folder
import { streamText, generateText, type CoreMessage } from '../node_modules/ai/dist/index.mjs';
import { getAgentConfig } from './registry';
import { agentTools, type AgentToolName } from './tools';

// Re-export types from core to maintain compatibility
export type { MessageWithParts } from '../core/state';

/**
 * Options for running an agent.
 */
export interface AgentOptions {
  /** Model ID (e.g. 'claude-sonnet-4-20250514') */
  modelId: string;
  /** Agent name from registry (e.g. 'build', 'explore', 'plan') */
  agentName: string;
  /** Working directory for file operations */
  workingDir: string;
  /** Optional abort signal for cancellation */
  abortSignal?: AbortSignal;
}

/**
 * Events emitted during agent streaming.
 */
export type StreamEvent =
  | { type: 'text'; data?: string }
  | { type: 'tool_call'; toolName?: string; toolId?: string; args?: unknown }
  | { type: 'tool_result'; toolId?: string; toolOutput?: string }
  | { type: 'error'; error?: string }
  | { type: 'done' };

/**
 * Get enabled tools for an agent based on its configuration.
 */
function getEnabledTools(agentName: string): Record<string, typeof agentTools[AgentToolName]> {
  const config = getAgentConfig(agentName);
  const enabledTools: Record<string, typeof agentTools[AgentToolName]> = {};

  for (const [name, tool] of Object.entries(agentTools)) {
    const enabled = config.toolsEnabled[name as AgentToolName];
    // Default to true if not explicitly disabled
    if (enabled !== false) {
      enabledTools[name] = tool;
    }
  }

  return enabledTools;
}

/**
 * Stream agent responses with tool execution.
 *
 * @example
 * ```typescript
 * for await (const event of streamAgent(messages, options)) {
 *   if (event.type === 'text') {
 *     process.stdout.write(event.data ?? '');
 *   }
 * }
 * ```
 */
export async function* streamAgent(
  messages: CoreMessage[],
  options: AgentOptions
): AsyncGenerator<StreamEvent, void, unknown> {
  const config = getAgentConfig(options.agentName);
  const enabledTools = getEnabledTools(options.agentName);
  const model = anthropic(options.modelId);

  try {
    const result = streamText({
      model,
      messages,
      system: config.systemPrompt,
      tools: enabledTools,
      abortSignal: options.abortSignal,
      maxSteps: 10, // Allow multi-step tool execution
      temperature: config.temperature,
      topP: config.topP,
    });

    for await (const part of (await result).fullStream) {
      switch (part.type) {
        case 'text-delta':
          yield { type: 'text', data: part.textDelta };
          break;

        case 'tool-call':
          yield {
            type: 'tool_call',
            toolName: part.toolName,
            toolId: part.toolCallId,
            args: part.args,
          };
          break;

        case 'tool-result':
          yield {
            type: 'tool_result',
            toolId: part.toolCallId,
            toolOutput: typeof part.result === 'string' ? part.result : JSON.stringify(part.result),
          };
          break;

        case 'error':
          yield { type: 'error', error: String(part.error) };
          break;

        case 'finish':
          yield { type: 'done' };
          break;
      }
    }
  } catch (error) {
    yield { type: 'error', error: error instanceof Error ? error.message : String(error) };
  }
}

/**
 * Run agent and return the complete response (non-streaming).
 */
export async function runAgent(
  messages: CoreMessage[],
  options: AgentOptions
): Promise<string> {
  const config = getAgentConfig(options.agentName);
  const enabledTools = getEnabledTools(options.agentName);
  const model = anthropic(options.modelId);

  const result = await generateText({
    model,
    messages,
    system: config.systemPrompt,
    tools: enabledTools,
    abortSignal: options.abortSignal,
    maxSteps: 10,
    temperature: config.temperature,
    topP: config.topP,
  });

  return result.text;
}

/**
 * Run agent synchronously (blocking wrapper for testing).
 *
 * Note: This is not truly synchronous - it blocks the event loop.
 * Use only for testing or CLI tools where blocking is acceptable.
 */
export function runAgentSync(
  messages: CoreMessage[],
  options: AgentOptions
): string {
  // Use Bun's synchronous promise resolution
  let result = '';
  let error: Error | null = null;
  let settled = false;
  const maxWaitMs = 300000; // 5 minute timeout
  const startTime = Date.now();

  runAgent(messages, options)
    .then((r) => { result = r; settled = true; })
    .catch((e) => { error = e; settled = true; });

  // Block until promise resolves (Bun-specific)
  // @ts-expect-error - Bun internal API
  if (typeof Bun !== 'undefined' && Bun.sleepSync) {
    while (!settled && (Date.now() - startTime) < maxWaitMs) {
      // @ts-expect-error - Bun internal API
      Bun.sleepSync(10);
    }

    if (!settled) {
      throw new Error('runAgentSync timed out after 5 minutes');
    }
  }

  if (error) throw error;
  return result;
}

/**
 * Stream agent with persistence to database.
 *
 * Wraps streamAgent and persists events using the database helpers.
 */
export async function* persistedStreamAgent(
  messages: CoreMessage[],
  options: AgentOptions,
  sessionId: string
): AsyncGenerator<StreamEvent, void, unknown> {
  // Import dynamically to avoid circular dependencies
  const { appendStreamingPart, updateMessageStatus, generateMessageId } = await import('../db/agent-state');

  // Generate a message ID at the start for consistent persistence
  const messageId = generateMessageId();
  let sortOrder = 0;

  try {
    for await (const event of streamAgent(messages, options)) {
      // Persist event to database
      if (event.type === 'text' && event.data) {
        await appendStreamingPart(sessionId, messageId, {
          type: 'text',
          text: event.data,
          sort_order: sortOrder++,
          time_start: Date.now(),
        });
      } else if (event.type === 'tool_call') {
        await appendStreamingPart(sessionId, messageId, {
          type: 'tool',
          tool_name: event.toolName,
          tool_state: {
            status: 'calling',
            toolId: event.toolId,
            args: event.args,
          },
          sort_order: sortOrder++,
          time_start: Date.now(),
        });
      } else if (event.type === 'tool_result') {
        await appendStreamingPart(sessionId, messageId, {
          type: 'tool',
          tool_name: 'result',
          tool_state: {
            status: 'complete',
            toolId: event.toolId,
            result: event.toolOutput,
          },
          sort_order: sortOrder++,
          time_start: Date.now(),
          time_end: Date.now(),
        });
      }

      yield event;

      if (event.type === 'done') {
        await updateMessageStatus(sessionId, messageId, 'complete');
      }
    }
  } catch (error) {
    await updateMessageStatus(sessionId, messageId, 'error');
    throw error;
  }
}

File: server/src/ai/agent.zig
const std = @import("std");
const client = @import("client.zig");
const types = @import("types.zig");
const registry = @import("registry.zig");
const tools_mod = @import("tools/mod.zig");

const StreamEvent = types.StreamEvent;
const AgentOptions = types.AgentOptions;
const StreamCallbacks = types.StreamCallbacks;
const ToolContext = types.ToolContext;

/// Run agent (non-streaming) and return final text
pub fn runAgent(
    allocator: std.mem.Allocator,
    messages: []const client.Message,
    options: AgentOptions,
    ctx: ToolContext,
) ![]const u8 {
    const config = registry.getAgentConfig(options.agent_name);

    // Get enabled tools for this agent
    const enabled_tools = try tools_mod.getEnabledTools(
        allocator,
        options.agent_name,
        config.tools_enabled,
    );
    defer allocator.free(enabled_tools);

    // Get API key
    const api_key = std.process.getEnvVarOwned(allocator, "ANTHROPIC_API_KEY") catch {
        return error.ApiKeyNotSet;
    };
    defer allocator.free(api_key);

    // Create client
    const anthropic = client.AnthropicClient.init(allocator, api_key);

    // Run agent loop
    var current_messages = std.ArrayList(client.Message){};
    defer current_messages.deinit(allocator);

    for (messages) |msg| {
        try current_messages.append(allocator, msg);
    }

    var steps: u32 = 0;
    const max_steps: u32 = 10;

    while (steps < max_steps) {
        const response = try anthropic.sendMessages(
            options.model_id,
            current_messages.items,
            config.system_prompt,
            enabled_tools,
            config.temperature,
            4096,
        );

        // Check for tool calls
        var has_tool_calls = false;
        var text_result: ?[]const u8 = null;

        for (response.content) |block| {
            switch (block) {
                .text => |text| {
                    text_result = text;
                },
                .tool_use => {
                    has_tool_calls = true;
                },
            }
        }

        if (!has_tool_calls) {
            // No tool calls, return text
            return text_result orelse error.NoContent;
        }

        // Process tool calls
        var assistant_parts = std.ArrayList(client.ContentPart){};
        var tool_results = std.ArrayList(client.ContentPart){};

        for (response.content) |block| {
            switch (block) {
                .text => |text| {
                    try assistant_parts.append(allocator, .{
                        .text = .{ .text = text },
                    });
                },
                .tool_use => |tu| {
                    try assistant_parts.append(allocator, .{
                        .tool_use = .{
                            .id = tu.id,
                            .name = tu.name,
                            .input = tu.input,
                        },
                    });

                    // Execute tool - parse JSON string to Value
                    const parsed_input = try std.json.parseFromSlice(std.json.Value, allocator, tu.input, .{});
                    defer parsed_input.deinit();

                    const output = tools_mod.executeTool(allocator, tu.name, parsed_input.value, ctx) catch |err| {
                        try tool_results.append(allocator, .{
                            .tool_result = .{
                                .tool_use_id = tu.id,
                                .content = @errorName(err),
                            },
                        });
                        continue;
                    };

                    try tool_results.append(allocator, .{
                        .tool_result = .{
                            .tool_use_id = tu.id,
                            .content = output,
                        },
                    });
                },
            }
        }

        // Add assistant message
        try current_messages.append(allocator, .{
            .role = .assistant,
            .content = .{ .parts = try assistant_parts.toOwnedSlice(allocator) },
        });

        // Add tool results as user message
        try current_messages.append(allocator, .{
            .role = .user,
            .content = .{ .parts = try tool_results.toOwnedSlice(allocator) },
        });

        steps += 1;
    }

    return error.MaxStepsReached;
}

/// Stream agent execution with callbacks
pub fn streamAgent(
    allocator: std.mem.Allocator,
    messages: []const client.Message,
    options: AgentOptions,
    ctx: ToolContext,
    callbacks: StreamCallbacks,
) !void {
    const config = registry.getAgentConfig(options.agent_name);

    // Get enabled tools for this agent
    const enabled_tools = try tools_mod.getEnabledTools(
        allocator,
        options.agent_name,
        config.tools_enabled,
    );
    defer allocator.free(enabled_tools);

    // Get API key
    const api_key = std.process.getEnvVarOwned(allocator, "ANTHROPIC_API_KEY") catch {
        callbacks.on_event(.{ .error_event = .{ .message = "ANTHROPIC_API_KEY not set" } }, callbacks.context);
        return;
    };
    defer allocator.free(api_key);

    // Create client
    const anthropic = client.AnthropicClient.init(allocator, api_key);

    // Run agent loop with streaming events
    var current_messages = std.ArrayList(client.Message){};
    defer current_messages.deinit(allocator);

    for (messages) |msg| {
        try current_messages.append(allocator, msg);
    }

    var steps: u32 = 0;
    const max_steps: u32 = 10;

    while (steps < max_steps) {
        const response = anthropic.sendMessages(
            options.model_id,
            current_messages.items,
            config.system_prompt,
            enabled_tools,
            config.temperature,
            4096,
        ) catch |err| {
            callbacks.on_event(.{ .error_event = .{ .message = @errorName(err) } }, callbacks.context);
            return;
        };

        // Emit events for response content
        var has_tool_calls = false;

        for (response.content) |block| {
            switch (block) {
                .text => |text| {
                    callbacks.on_event(.{ .text = .{ .data = text } }, callbacks.context);
                },
                .tool_use => |tu| {
                    has_tool_calls = true;
                    callbacks.on_event(.{
                        .tool_call = .{
                            .tool_name = tu.name,
                            .tool_id = tu.id,
                            .args = tu.input,
                        },
                    }, callbacks.context);
                },
            }
        }

        if (!has_tool_calls) {
            // Done
            callbacks.on_event(.{ .done = {} }, callbacks.context);
            return;
        }

        // Process tool calls
        var assistant_parts = std.ArrayList(client.ContentPart){};
        var tool_results = std.ArrayList(client.ContentPart){};

        for (response.content) |block| {
            switch (block) {
                .text => |text| {
                    try assistant_parts.append(allocator, .{
                        .text = .{ .text = text },
                    });
                },
                .tool_use => |tu| {
                    try assistant_parts.append(allocator, .{
                        .tool_use = .{
                            .id = tu.id,
                            .name = tu.name,
                            .input = tu.input,
                        },
                    });

                    // Execute tool - parse JSON string to Value
                    const parsed_input = try std.json.parseFromSlice(std.json.Value, allocator, tu.input, .{});
                    defer parsed_input.deinit();

                    const output = tools_mod.executeTool(allocator, tu.name, parsed_input.value, ctx) catch |err| {
                        callbacks.on_event(.{
                            .tool_result = .{
                                .tool_id = tu.id,
                                .tool_output = @errorName(err),
                            },
                        }, callbacks.context);

                        try tool_results.append(allocator, .{
                            .tool_result = .{
                                .tool_use_id = tu.id,
                                .content = @errorName(err),
                            },
                        });
                        continue;
                    };

                    callbacks.on_event(.{
                        .tool_result = .{
                            .tool_id = tu.id,
                            .tool_output = output,
                        },
                    }, callbacks.context);

                    try tool_results.append(allocator, .{
                        .tool_result = .{
                            .tool_use_id = tu.id,
                            .content = output,
                        },
                    });
                },
            }
        }

        // Add assistant message
        try current_messages.append(allocator, .{
            .role = .assistant,
            .content = .{ .parts = try assistant_parts.toOwnedSlice(allocator) },
        });

        // Add tool results as user message
        try current_messages.append(allocator, .{
            .role = .user,
            .content = .{ .parts = try tool_results.toOwnedSlice(allocator) },
        });

        steps += 1;
    }

    callbacks.on_event(.{ .error_event = .{ .message = "Max steps reached" } }, callbacks.context);
}

test "registry integration" {
    const config = registry.getAgentConfig("build");
    try std.testing.expectEqualStrings("build", config.name);
}
